src/pyodide/internal/snapshot.js

```diff
@@ -0,0 +1,414 @@
+import { default as ArtifactBundler } from "pyodide-internal:artifacts";
+import { default as UnsafeEval } from "internal:unsafe-eval";
+import {
+  SITE_PACKAGES_INFO,
+  SITE_PACKAGES_SO_FILES,
+  getSitePackagesPath,
+} from "pyodide-internal:setupPackages";
+import { default as TarReader } from "pyodide-internal:packages_tar_reader";
+import processScriptImports from "pyodide-internal:process_script_imports.py";
+import {
+  IS_CREATING_BASELINE_SNAPSHOT,
+  MEMORY_SNAPSHOT_READER,
+} from "pyodide-internal:metadata";
+import { reportError, simpleRunPython } from "pyodide-internal:util";
+
+/**
+ * This file is a simplified version of the Pyodide loader:
+ * https://github.com/pyodide/pyodide/blob/main/src/js/pyodide.ts
+ *
+ * In particular, it drops the package lock, which disables
+ * `pyodide.loadPackage`. In trade we add memory snapshots here.
+ */
+
+/**
+ * _createPyodideModule and pyodideWasmModule together are produced by the
+ * Emscripten linker
+ */
+import { _createPyodideModule } from "pyodide-internal:generated/pyodide.asm";
+
+const SHOULD_UPLOAD_SNAPSHOT =
+  ArtifactBundler.isEnabled() || ArtifactBundler.isEwValidating();
+
+/**
+ * Global variable for the memory snapshot. On the first run we stick a copy of
+ * the linear memory here, on subsequent runs we can skip bootstrapping Python
+ * which is quite slow. Startup with snapshot is 3-5 times faster than without
+ * it.
+ */
+let READ_MEMORY = undefined;
+let SNAPSHOT_SIZE = undefined;
+export let SHOULD_RESTORE_SNAPSHOT = false;
```

Comment:
Agreed, but I think fixing this would best be left to a followup; I'm trying to avoid changes if at all possible here b/c moving all the code makes the diff hard to read.